frontend/src/pages/queryBuilder/QueryGraph.jsx

```diff
@@ -0,0 +1,297 @@
+/* eslint-disable indent, no-use-before-define, func-names, no-return-assign */
+import React, {
+  useEffect, useContext, useRef, useMemo,
+} from 'react';
+import * as d3 from 'd3';
+import graphUtils from '~/utils/d3/graph';
+import nodeUtils from '~/utils/d3/nodes';
+import edgeUtils from '~/utils/d3/edges';
+import highlighter from '~/utils/d3/highlighter';
+
+import BiolinkContext from '~/context/biolink';
+import QueryBuilderContext from '~/context/queryBuilder';
+import getNodeCategoryColorMap from '~/utils/colors';
+import queryGraphUtils from '~/utils/queryGraph';
+
+const nodeRadius = 40;
+
+/**
+ * Main D3 query graph component
+ * @param {int} height - height of the query graph element
+ * @param {int} width - width of the query graph element
+ * @param {obj} clickState - dict of graph click state properties
+ * @param {func} updateClickState - reducer to update graph click state
+ */
+export default function QueryGraph({
```

Comment:
The name `QueryGraph` is a bit weird because it can refer to the part of the message as well as this editor. I'm not sure how big of a change this is but maybe we could call this component a VisualEditor? That works nicely with the TextEditor component.

Reply:
I see where you're coming from, but I don't know if there's a more descriptive name than QueryGraph. Maybe QueryGraphDisplay? In my mind, VisualEditor could refer to the entire builder, as text is also "visual". I'm open to other suggestions.